extra/import_data.py

```diff
@@ -0,0 +1,174 @@
+            logger.info('removed: %s from database, start replacing with newer version' % patient_dn_no)
+        elif last_updated_on_Alissa < last_updated_on_mongoDB:
+            logger.info('lastUpdatedOn older than within the database for %s, this should not happen' % patient_dn_no)
+            # TODO send (email) notification of this error, see issue #11 GitHub
+
+    # get all relevant info from one patient into one dictionary.
+    patient = {}
+    patient["dn_no"] = patient_dn_no
+    logger.info('start uploading to mongodb: %s' % patient_dn_no)
+    patient["patient_accession_no"] = accession_number
+    patient.update(inheritance_analysis)
+    patient["analysis"] = patient.pop("reference")
+    patient["platform_dataset"] = analyis_sources["platformDataSet"]["info"]
+
+    # changes the format of the annotation sources from Alissa, to fit in MongoDB and resemble the old O-schijf format
+    externalSources_dict = {}
+    for source in analyis_sources['annotationSources']:
+        source_name = ""
+        source_value = ""
+        for key, value in source.items():
+            if key == 'name':
+                source_name = value
+            elif key == 'description':
+                source_value = value
+        if source_name and source_value:
+            externalSources_dict[source_name] = source_value
+        else:
+            logger.info("error, externalsources has changed format in Alissa")
+    patient["annotation_sources"] = externalSources_dict  # add previous section to total patient info
+
+    # extract information from the VUS/variant data and add to "patient"
+    for variant in vus_export:
+        # format for gnomad Links. 4 availble in Alissa (snp, insertion, deletion and substitution)
+        try:
+            fullgnomen = variant['platformDatasets']['HGVS genomic-level nomenclature (fullGNomen)']
+            variant['fullgnomen'] = fullgnomen  # NC_000001.10:g.12345678T>A
+            if fullgnomen:  # NC_000001.10:g.12345678T>A
+                gnomad_data = re.split(':[a-z].', fullgnomen)[1]  # 123456789T>A
+                if variant["type"] == "snp":
+                    gnomad_data = [c for c in re.split(r'([-+]?\d*\.\d+|\d+)', gnomad_data) if c]  # 123456789T>A
+                    gnomad_data = variant["chromosome"] + "-" + re.sub('[<>]+', '-', ("-".join(gnomad_data)))  # 1-12345678-T-A
+                    variant['GnomadVariant'] = {'Single nucleotide variant': gnomad_data}
+                elif variant["type"] in ["insertion", "deletion", "substitution"]:
+                    gnomad_data = variant["chromosome"] + "-" + gnomad_data
+                    variant['GnomadVariant'] = {variant["type"].capitalize(): gnomad_data}
+                    # TODO: make link format correctly for "insertion", "deletion", "substitution" genomic variations
+            else:  # on rare occasions, fullGNomen is empty
+                variant['GnomadVariant'] = {variant["type"]: ''}
```

Comment:
In my opinion, this try block is too big and not specific enough. Is the keyError and TypeError likely to happen in each line?

Reply:
yes and both are due to incomplete entries in Alissa. So there is "technically" a VUS, but it is a very old entry that does not actually contain any data hence the error message.

Reply:
Is it possible to check whether data is complete before extracting and checking all other conditions? In that way the try becomes more specific and as a benefit also smaller.

Reply:
there are several checks in place to check for complete data. most of them (should be all) can be checked if retrieved from Alissa. However, this error only occurred ~5 times on the whole Alissa database. The person in charge already made a ticket about these weird entries in Alissa.

So technically, yes I could check beforehand, but then I should check every key-value in Alissa (which I think is not feasable)

Reply:
We talked about this, and one of the issues is also a too long try-except.
would this be a solution?
```
    for variant in vus_export:
        # format for gnomad Links. 4 availble in Alissa (snp, insertion, deletion and substitution)
        try:
            fullgnomen = variant['platformDatasets']['HGVS genomic-level nomenclature (fullGNomen)']
            variant['fullgnomen'] = fullgnomen  # NC_000001.10:g.12345678T>A
        except (KeyError, TypeError):
            logger.error('Variant in patient %s has no platformDatasets and fullGNomen, variant not uploaded' % patient_dn_no)
            continue
        if fullgnomen:  # NC_000001.10:g.12345678T>A
            gnomad_data = re.split(':[a-z].', fullgnomen)[1]  # 123456789T>A
            if variant["type"] == "snp":
            ......
```

Reply:
I have tested above solution without "typeError" to see what exactly causes this error (which line mainly).

```
2022-06-03 20:35:08,769 [INFO] __main__ - start Alissa retrieval of: DN201803750 with analysisID: 33569
2022-06-03 20:35:19,333 [INFO] __main__ - Alissa retrieval completed of: DN201803750
2022-06-03 20:35:19,342 [INFO] __main__ - start uploading to mongodb: DN201803750
Traceback (most recent call last):
  File "import_data.py", line 169, in <module>
    import_from_alissa(alissa_client, start_time, logger)
  File "import_data.py", line 69, in import_from_alissa
    upload_to_mongodb(inheritance_analysis, accession_number, analyis_sources, patient_dn_no, vus_export, logger)
  File "import_data.py", line 128, in upload_to_mongodb
    fullgnomen = variant['platformDatasets']['HGVS genomic-level nomenclature (fullGNomen)']
TypeError: string indices must be integers
```

So I think that if the key is wrong/non-existant you get the "keyError", for the "typeError" it happens at the same line, when the index is non-existant [stackoverflowexample](https://stackoverflow.com/questions/6077675/why-am-i-seeing-typeerror-string-indices-must-be-integers)

The variable item is a string. An index looks like this:
```
mystring = 'helloworld'
print mystring[0]
'h'
```
The above example uses the 0 index of the string to refer to the first character.
Strings can't have string indices (like dictionaries can). So this won't work:
```
mystring = 'helloworld'
print mystring['stringindex']
TypeError: string indices must be integers
```

So in this case, I think my code example will work for both the typeError and the keyError

Reply:
Right. You could split this into two try-except, one based on `variant['platformDatasets']` and another on the second key. However, might be my personal preference.